cpp/oneapi/dal/test/global_context.cpp

```diff
@@ -0,0 +1,148 @@
+/*******************************************************************************
+* Copyright 2024 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/test/engine/common.hpp"
+#include "oneapi/dal/detail/global_context.hpp"
+
+namespace oneapi::dal::detail::test {
+
+#if defined(__x86_64__)
+
+/// x86 CPU registers relevant to CPUID instruction
+struct cpuid_registers {
+    uint32_t eax_;
+    uint32_t ebx_;
+    uint32_t ecx_;
+    uint32_t edx_;
+};
+
+/// Runs CPUID instruction
+///
+/// @param[in]  eax       The input value of EAX register, the ID of CPUID subfunction.
+///                       By that value CPUID defines which information should be returned.
+///                       For example:
+///                           eax == 0 - CPU vendor ID is returned;
+///                           eax == 1 - Processor type, family, stepping, ... are returned;
+///                           ...
+/// @param[out] registers x86 CPU registers to write the results of CPUID
+/// @param[in]  ecx       The input value of ECX register, the ID of CPUID's subfunction subleaf.
+static inline void cpuid(uint32_t eax, cpuid_registers& registers, uint32_t ecx = 0U) {
+    asm volatile(
+        "cpuid"
+        : "=a"(registers.eax_), "=b"(registers.ebx_), "=c"(registers.ecx_), "=d"(registers.edx_)
+        : "a"(eax), "c"(ecx));
+}
+
+/// Copies the value of x86 register into the buffer of characters
+///
+/// @param[in]  reg    The x86 register value
+/// @param[out] buffer Pointer to the output character buffer
+void register_to_buffer(uint32_t reg, char* buffer) {
+    ONEDAL_ASSERT(buffer);
+    constexpr uint32_t mask = 0xFF; // lower 8 bits
+    constexpr uint32_t register_size = sizeof(uint32_t);
+    for (uint32_t i = 0; i < register_size; ++i) {
```

Comment:
Looks like little-endian write of bytes, so you don't need the copy like this? The uint32_t are little endian on x86, so you can just write out the value. Assuming 8-bit char, of course
```
ONEDAL_ASSERT(buffer);
*(uint32_t*)buffer = reg;
```